File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;
class Solution {
    public int solution(int cacheSize, String[] cities) {
        Map<String,Integer> map = new HashMap<>(); // 호출되는 카운트
        List<String> list = new ArrayList<>(); // key 
        
        int answer = 0;
        int idx = 0;
        for(String city: cities){
            idx++;
            city = city.toLowerCase();
            if(!map.containsKey(city)){
                map.put(city,idx);
                list.add(city);
                answer+=5;
            }
            else{
                boolean check = false;
                for(int i=0; i<cacheSize ; i++){
                    if(city.equals(list.get(i))){
                        check = true;
                        break;
                    }
                }
                answer += check ? 1:5;
                map.replace(city, idx);    
            }
            //정렬
            Collections.sort(list, (o1,o2) -> Integer.compare(map.get(o2),map.get(o1)));
            
        }
        
        return answer;
    }
}

File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque

def solution(cacheSize, cities):
    if cacheSize == 0:
        return len(cities)*5;
    q = deque()
    answer = 0 
    for idx,city in enumerate(cities):
        city = city.upper()
        if city in q:
            answer += 1
            q.remove(city)
        else:
            answer+=5
            if len(q) >= cacheSize:
                q.popleft()
        q.append(city)
    return answer